keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1550 @@
+
+This proposal splits the full implementation of QoS-class resources into
+multiple phases, building functionality gradually, step-by-step. The goal is to
+make the discussions more focused and easier. We may also learn on the way,
+insights from earlier phases affecting design choises made in the later phases,
+hopefully resulting in a better overall end result. However, we also outline
+all the future steps to not lose the overall big picture.
+
+In the current design QoS-class resources are designed to be opaque to the CRI
+client in the sense that the container runtime takes care of configuration and
+control of the resources and the classes within.
+
+### Phase 1
+
+This KEP (the [Proposal](#proposal)) implements the first phase. The goal is to
+enable a bare minimum for users to leverage QoS-class resources and start
+experimenting with them in Kubernetes:
+
+- extend the CRI protocol to allow QoS-class resource assignment and updates
+- implement pod annotations as an initial user interface
+- introduce a feature gate for enabling QoS-class resource support in kubelet
+
+### Future work
+
+This section sheds light on the end goal of this work in order to better
+evaluate this KEP in a broader context. What a fully working solution would
+consists of and what the (next) steps to accomplish that would be. These topics
+are currently out of the scope of this KEP and were listed under
+[Non-goals](#non-goals).
+
+#### Pod Spec
+
+Replace pod annotations with proper user interface via the Pod spec. Below, one
+possible option is presented.
+
+Introduce a new field (e.g. class) into ResourceRequirements of Container.
+
+```diff
+// ResourceRequirements describes the compute resource requirements.
+type ResourceRequirements struct {
+     // Limits describes the maximum amount of compute resources allowed.
+     Limits ResourceList `json:"limits,omitempty"
+     // Requests describes the minimum amount of compute resources required.
+     Requests ResourceList `json:"requests,omitempty"
++    // Classes specifies the QoS-class resources that the container should be assigned
++    Classes map[ClassResourceName]string
+}
+
++// ClassResourceName is the name of a QoS-class resource.
++type ClassResourceName string
+```
+
+Also, we add a `Resources` field to the `PodSpec`. We will re-use the existing
+`ResourceRequirements` type but Limits and Requests must be left empty. Classes
+may be set and they represent the Pod-level assignment of QoS-class resources,
+comparable to the PodClassResources message in PodSandboxConfig in the CRI API.
+
+```diff
+ type PodSpec struct {
+@@ -224,4 +224,8 @@ type PodSpec struct {
+     // Default to false.
+     // +optional
+     SetHostnameAsFQDN *bool `json:"setHostnameAsFQDN,omitempty" protobuf:"varint,35,opt,name=setHostnameAsFQDN"`
++    // Pod-level resources. Currently, requests and limits are not allowed
++    // to be specified for pods.
++    // +optional
++    Resources ResourceRequirements
```

Comment:
something that's tough here is this blurs the line between ownership of the pod cgroup between kubelet and cri. Historically, kubelet was responsible for the pod (because docker had no notion of pods). I don't think there's any way to fix this, but maybe it should be mentioned or called out?

Reply:
Mm, I get that. This KEP is not aiming to change that division (even though that might make sense, thinking e.g. about vm-based runtimes, but that's another story 😊). We have PodSandboxConfig in cri-api, and even resources fields in the LinuxPodSandboxConfig so it seems to get blurrier one step at a time... Fortunately, the end user (writer of the PodSpec) does not need to know the technical details where this is implemented/enforced, even though it's of course important on our side.


[KEP-2837](https://github.com/kubernetes/enhancements/pull/1592) that is referenced below (link was broken but now fixed) is aiming to introduce a similar pod-level resources field. So these two KEPs are sort of aligned on this front.

I didn't change anything here now but could add a note e.g.:

> NOTE: Historically, kubelet has been responsible for the pod-level resources (cgroups) and container runtime has owned the container-level resources. This KEP is not directly changing that division

Or maybe something like that^ should be added to the cri-api proposal part? WDYT?

Reply:
resolved, as there is no pod-level `Resources` field anymore (has been replaced with `QOSResources` field)